File: 270/find_closest_value_bst.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None
import sys

class Solution(object):
	def closestValue(self, root, target):
		"""
		:type root: TreeNode
		:type target: float
		:rtype: int
		"""
		if root == None:
			return None

		if root.left == None and root.right == None:
			return root.val

		if target < root.val:
			if root.left == None:
				return root.val

			max_left = self.get_max(root.left)
			if target > max_left:
				if self.abs_diff(max_left, target) < self.abs_diff(root.val, target):
					return self.closestValue(root.left, target)
				return root.val

			return self.closestValue(root.left, target)

		if target > root.val:
			if root.right == None:
				return root.val

			min_right = self.get_min(root.right)
			if target < min_right:
				if self.abs_diff(min_right, target) < self.abs_diff(root.val, target):
					return self.closestValue(root.right, target)
				return root.val

			return self.closestValue(root.right, target)

		else:
			return root.val

	def get_max(self, root):
		if root.right != None:
			return self.get_max(root.right)
		return root.val

	def get_min(self, root):
		if root.left != None:
			return self.get_min(root.left)
		return root.val

	def abs_diff(self, val1, val2):
		return abs(val1 - val2)

	def get_least_diff(self, target, ls_val):
		ls_diff = [abs(target - val) for val in ls_val]
		return ls_val[ls_diff.index(min(ls_diff))]

File: 270/FindClosestValueBST.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class FindClosestValueBST {
	// 2ms
    public int closestValue(TreeNode root, double target) {
    	return __closestValue__(root, target);
    }

    int getMax(TreeNode root) {
    	assert root!=null;

    	while (root.right != null)
    		root = root.right;
    	return root.val;
    }

    int getMin(TreeNode root) {
    	assert root!=null;

    	while (root.left != null)
    		root = root.left;
    	return root.val;
    }

    // include parent Value
    public int __closestValue__(TreeNode root, double target) {
    	assert root!=null;

    	if (target <= (float)root.val) {
    		if (root.left != null) {
    			int leftmax = getMax(root.left);
    			if (target > (double) leftmax) {
    				// compare max node from left with current node
					double leftmaxdist = target - leftmax;
					double rootdist = root.val - target;
    				if (leftmaxdist < rootdist) {
    					return leftmax;
    				}
    				return root.val;
    			}
    			return __closestValue__(root.left, target);
    		}
    		return root.val;
    	}

    	else {
    		if (root.right != null) {
    			int rightmin = getMin(root.right);
    			if (target < (double) rightmin) {
    				double rightmindist = rightmin - target;
    				double rootdist = target - root.val;
    				if (rightmindist < rootdist) {
    					return rightmin;
    				}
    				return root.val;
    			}
    			return __closestValue__(root.right, target);
    		}
    		return root.val;
    	}
    }

    public static void main(String[] args) {
    	System.out.println("Hello World!");
    	return;
    }
}